File: apps/python/event_sourcing/restocker.py
"""Restocker aggregate and command handler implementation."""

from abc import ABC, abstractmethod
from typing import List

from .command import COMMANDS, RestockOrder
from .event import EVENTS, RestockOrdered
from .command_handler import CommandHandler


class RestockCommandHandler(CommandHandler):
    """Command handler for restock commands."""

    def __init__(self, events: List[EVENTS]):
        self._history = events
        self._publish: List[EVENTS] = []

    def handle(self, command: COMMANDS) -> List[EVENTS]:
        """Handle a restock command."""
        if isinstance(command, RestockOrder):
            state = RestockerState(self._history)
            restocker = Restocker(state)
            events = restocker.handle(command)
            self._publish = events
            return events
        return []


class Aggregate(ABC):
    """Base class for aggregates."""

    @abstractmethod
    def handle(self, command: COMMANDS) -> List[EVENTS]:
        """Handle a command and return emitted events."""
        pass


class RestockerState:
    """State rebuilt from restock events."""

    def __init__(self, events: List[EVENTS]):
        self.quantity = 0

        for event in events:
            if event.type == RestockOrdered.TYPE:
                self.quantity += event.quantity


class Restocker(Aggregate):
    """Aggregate that handles restock commands."""

    def __init__(self, state: RestockerState):
        self.state = state

    def handle(self, command: COMMANDS) -> List[EVENTS]:
        """Handle a command and return emitted events."""
        if isinstance(command, RestockOrder):
            if self.state.quantity < 100:
                return [RestockOrdered(command.quantity)]
        return []

File: apps/rust/src/restocker.rs
//! Restocker aggregate and command handler implementation.

use crate::command::{Command, RestockOrder};
use crate::command_handler::CommandHandler;
use crate::event::{Event, RestockOrdered};

/// Command handler for restock commands.
pub struct RestockCommandHandler {
    history: Vec<RestockOrdered>,
    publish: Vec<RestockOrdered>,
}

impl RestockCommandHandler {
    /// Create a new command handler with the given event history.
    pub fn new(history: Vec<RestockOrdered>) -> Self {
        Self {
            history,
            publish: Vec::new(),
        }
    }

    /// Get the published events.
    pub fn published(&self) -> &[RestockOrdered] {
        &self.publish
    }
}

impl CommandHandler<RestockOrder, RestockOrdered> for RestockCommandHandler {
    fn handle(&mut self, command: &RestockOrder) -> Vec<RestockOrdered> {
        let state = RestockerState::new(&self.history);
        let mut restocker = Restocker::new(state);
        let events = restocker.handle(command);
        self.publish = events.clone();
        events
    }
}

/// Aggregate trait for handling commands.
pub trait Aggregate<C: Command, E: Event> {
    /// Handle a command and return emitted events.
    fn handle(&mut self, command: &C) -> Vec<E>;
}

/// State rebuilt from restock events.
#[derive(Debug, Clone)]
pub struct RestockerState {
    pub quantity: i32,
}

impl RestockerState {
    /// Create state by replaying events.
    pub fn new(events: &[RestockOrdered]) -> Self {
        let mut quantity = 0;

        for event in events {
            if event.event_type() == RestockOrdered::TYPE {
                quantity += event.quantity;
            }
        }

        Self { quantity }
    }
}

/// Aggregate that handles restock commands.
pub struct Restocker {
    state: RestockerState,
}

impl Restocker {
    pub fn new(state: RestockerState) -> Self {
        Self { state }
    }
}

impl Aggregate<RestockOrder, RestockOrdered> for Restocker {
    fn handle(&mut self, command: &RestockOrder) -> Vec<RestockOrdered> {
        if self.state.quantity < 100 {
            vec![RestockOrdered::new(command.quantity)]
        } else {
            vec![]
        }
    }
}
